pgxn/neon_walredo/shmpipe.c

```diff
@@ -270,6 +280,16 @@ void shmem_pipe_process_request(pipe_t* pipe, char const* req, size_t req_size,
 			do
 			{
 				/* wait until head is advanced */
+				if (++busy_loop_iterations % CHECK_WATCHDOG_INTERVAL == 0 && pipe->child_pid)
+				{
+					int wstatus;
+					if (waitpid(pipe->child_pid, &wstatus, WNOHANG) > 0)
+					{
+						/* Child process is terminated */
+						shmem_pipe_reset(pipe);
```

Comment:
Regarding this reset, how is it guaranteed that all producers coming after the thread doing the reset will be guaranteed to `return 0;` through this same or some different path, and will never wait a response?

A generational id or a counter for all `shmem_pipe_reset`'s could be added, read in the beginning of `shmem_pipe_process_request` and could be compared to during the producer phase. Unsure where or when yet.

Reply:
With the microbenchmarks `short/short/16` case (1kB payload) and `kill -9` of the `postgres --wal-redo` process, I managed to get a version where 10 out of 16 threads are busywaiting on `sched_yield`.

Reply:
> how is it guaranteed that all producers coming after the thread doing the reset will be guaranteed to return 0

No such warranty.

> I managed to get a version where 10 out of 16 threads are busy waiting on sched_yield. 

Yes, `waitpid` reslt should be compared with zero. If it returns -1 (process with such pid not found because it is already terminated and another `waitpid`  grab its's status), then it should be treated in the same way as that process is if child terminated.

Reply:
With commit 215e7cc3cf5643a6945a19dd856b342d2af9cec6 it's the same behaviour. I think it must be because of `shmem_pipe_reset` will overwrite the child_pid as zero, so the waiters will no longer try to get `waitpid == -1` and are instead just spinning.

I wonder if it should be the process restart which should handle:
1. the unblocking of all past producers
2. reset the structure
3. let all producers retry?

Reply:
Yes, agree with you.
Added "generation" field to the pipe to address this issue.

Reply:
This latest batch of commits looks good.